### PR TITLE
feat: add multimodal driver

### PR DESCRIPTION
**Pull request by bwanglzu**
**Changes introduced**
Added a new `MultiModalDriver` and a new executors interface `BaseMultiModalEncoder`.

- **MultiModalEncoder** has an `interface` encode that is a variable length argument of different batches of data. One `np.ndarray` per modality so that the executor can work in different ways.

- **MultiModalEncoder** has an argument that informs the `driver` which modality should be assigned to which index of the `variable length` argument.

- **MultiModalDriver** works on `chunks of documents` assuming that every `chunk` is from a different `modality` (otherwise it is an invalid document). So it assigns the `resulting embedding` to the `parent` document of the chunks with multiple modalities

- **MultiModalDriver** extracts from chunks content based on this logic: If it has an embedding, extract it, otherwise extract one of the actual content.

**Notes**
The traversal strategy looks quite similar as `Chunk2DocRankerDriver` but with `chunks`. This is related to the need to provide batching for `encoders`. Otherwise we may not be able to provide an efficient implementation for these executors.

**Next steps**
Work on the batching decorator to enable its capabilities to batch multiple inputs in a "zipped" way.

**Usage**
If modalities come from different `Flow` paths this driver needs to be combined with `ReduceAllDriver` in front of it. Just to showcase we have implemented for testing some `concatenation` executors. We could think on removing ConcatenateDriver because it also has too much coupling with `ReduceAllDriver`

### CODE REVIEW

**Comment by JoanFM** [comment]
It looks like the idea is correct, let's add tests to prove it

File: jina/drivers/multimodal.py
Comment (JoanFM) [inline jina/drivers/multimodal.py:74]: wrong method name

**Comment by JoanFM** [comment]
After we have some tests, we can see how we decide which part of the document (content or embedding) must be extracted per modality

**Comment by bwanglzu** [comment]
ok, while writing test I figured out a problem in the design. For each document, I grouped embeddings based on the same modality, and assign the concatenated embedding to `context_doc`. This means for each modality, I'm replacing the embedding with another modality embedding.

For example, we have 4 chunks with 2 modalities: 1 (visual), 2 (visual), 3 (textual), 4 (textual): embedding 1 and 2 were concatenated into embedding e1 and e1 was assigned as the embedding for `context_doc`. embedding3 and 4 were concatenated into embedding e2, and again, it was assigned as embedding for `context_doc`.

**Comment by JoanFM** [comment]
> ok, while writing test I figured out a problem in the design. For each document, I grouped embeddings based on the same modality, and assign the concatenated embedding to `context_doc`. This means for each modality, I'm replacing the embedding with another modality embedding.
> 
> For example, we have 4 chunks with 2 modalities: 1 (visual), 2 (visual), 3 (textual), 4 (textual): embedding 1 and 2 were concatenated into embedding `e1` and `e1` was assigned as the embedding for `context_doc`. embedding3 and 4 were concatenated into embedding `e2`, and again, it was assigned as embedding for `context_doc`. and we lose `e1`..

In this case (only remember that is a temporary thing to have concatenation), u should be able to concatenate all the embeddings of every modality.

Actually I would assume that per modality there is only one chunk at this point.

**Review by JoanFM — COMMENTED** [review]
I think in the scope of this PR and before merging, we should decide the interface of a Base class MultiModalEmbeddingExecutor so that we can have a test with a mock that does concatenation. To be done later

**Comment by bwanglzu** [comment]
@JoanFM thanks for the input, while then how can we differentiate `MultimodalDriver` and `ReduceAllDriver`?

**Comment by JoanFM** [comment]
> @JoanFM thanks for the input, while then how can we differentiate `MultimodalDriver` and `ReduceAllDriver`?

I do not quite understand the question, what is the issue?

**Comment by bwanglzu** [comment]
@JoanFM my brain stopped running :) ignore my previous question

File: jina/drivers/multimodal.py
Comment (maximilianwerk) [inline jina/drivers/multimodal.py:60]: This can be made easier with:
```
from collections import defaultdict
...
modal_doc = defaultdict(list)
for doc in docs:
    modal_doc[doc.modality].append(doc.id)
```

File: jina/drivers/multimodal.py
Comment (maximilianwerk) [inline jina/drivers/multimodal.py:80]: same here. `doc_pointers` can be `defaultdict(list)`

File: jina/drivers/multimodal.py
Comment (maximilianwerk) [inline jina/drivers/multimodal.py:62]: Perhaps someone knows better than me: In the past python dictionaries where not order preserving. So when you iterate a dictionary, you can get the items in random order. If this is true, the concatenated embedding is not deterministic. I guess that is bad, right?

Anyhow, I am not sure, whether this is still true for current python version.

**Review by maximilianwerk — COMMENTED** [review]
Some comments for cleaner code

File: jina/drivers/multimodal.py
Comment (bwanglzu) [reply jina/drivers/multimodal.py:60]: wow nice! thanks @maximilianwerk ! learned something new

File: jina/drivers/multimodal.py
Comment (bwanglzu) [reply jina/drivers/multimodal.py]: I'll leave the check here since `doc_pointer` is a parameter passed into the function.

File: jina/drivers/multimodal.py
Comment (bwanglzu) [reply jina/drivers/multimodal.py]: thanks for the comment, you're right, that's why we have `OrderedDict`. While concatenating is a temporary solution for this design (to be discussed further), so it's not an issue at the moment.

Comment (JoanFM) [reply jina/drivers/multimodal.py]: Yes concatrnation is a placeholder that needs to be changed by the call to the executor

**Review by JoanFM — COMMENTED** [review]
Are we here considering that there is only one input per modality?

**Comment by bwanglzu** [comment]
@JoanFM you mean one chunk per modality?

**Comment by JoanFM** [comment]
> @JoanFM you mean one chunk per modality?

Yes, does it make sense this assumption?

**Comment by bwanglzu** [comment]
@JoanFM yes it make sense (for most cases, e.g. one pic + text)

**Comment by JoanFM** [comment]
> @JoanFM yes it make sense (for most cases, e.g. one pic + text)

Yes I think we can assume, every executor would expect a single input per modality. If it has a need to merge two modalities before we can expect it to do in a previous driver

**Comment by JoanFM** [comment]
Since merging #1112 , I feel we should not have this driver inherit from ReduceDriver but have it a regular driver and assume that docs or chunks with different modalities come from different paths, then you need to add a ReduceDriver before it but if it comes from a previous crafter then you do not need it.

File: tests/unit/drivers/test_multimodal_driver.py
Comment (JoanFM) [inline tests/unit/drivers/test_multimodal_driver.py:64]: We decided that the `driver` cannot handle 2 chunks of the same modality. They should be handled by an upstream Pod  to be merged into one

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/drivers/multimodal.py
Comment (JoanFM) [inline jina/drivers/multimodal.py:36]: Look at the documentation here. We should validate this and raise an `Exception` if not fullfilled.

File: jina/drivers/multimodal.py
Comment (JoanFM) [inline jina/drivers/multimodal.py:67]: We assume one `chunk` per modality, so no point to do append, if there is a duplicated one, raise `Error`. Otherwise, the interface for the `executor` will turn quite messy.

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/unit/drivers/test_multimodal_driver.py
Comment (JoanFM) [inline tests/unit/drivers/test_multimodal_driver.py:33]: this is an invalid `document`. so good to test for an exception raised,

**Comment by JoanFM** [comment]
@bwanglzu there is some problem with this branch, I think some rebase has gone wrong.

**Comment by JoanFM** [comment]
@bwanglzu, I hope it is okey, I edited the description to explain in more detail the changes we have made

File: jina/drivers/multimodal.py
Comment (bhavsarpratik) [inline jina/drivers/multimodal.py:5]: Import sequence can be standard

File: jina/executors/encoders/multimodal/__init__.py
Comment (bhavsarpratik) [inline jina/executors/encoders/multimodal/__init__.py:6]: Import sequence can be standard

File: tests/integration/multimodal/test_multimodal_parallel.py
Comment (bhavsarpratik) [inline tests/integration/multimodal/test_multimodal_parallel.py:2]: Import sequence can be standard

File: jina/drivers/multimodal.py
Comment (bhavsarpratik) [inline jina/drivers/multimodal.py:74]: input_args = [content_by_modality[modality] for modality in self.position_by_modality.keys()]

**Review by bhavsarpratik — CHANGES_REQUESTED** [review]

File: jina/drivers/multimodal.py
Comment (JoanFM) [reply jina/drivers/multimodal.py:5]: what is the standard?

File: jina/drivers/multimodal.py
Comment (JoanFM) [reply jina/drivers/multimodal.py]: Done! Thanks!

File: jina/drivers/multimodal.py
Comment (bhavsarpratik) [reply jina/drivers/multimodal.py:5]: https://stackoverflow.com/questions/22722976/import-order-coding-standard

**Comment by JoanFM** [comment]
Can you check again @bhavsarpratik

**Review by bhavsarpratik — APPROVED** [review]

File: jina/drivers/multimodal.py
Comment (nan-wang) [inline jina/drivers/multimodal.py:72]: The position information is not used. 
```suggestion
        return [v for k, v in sorted(self.position_by_modality.items(), key=lambda item: item[1])]
```

Comment (nan-wang) [reply jina/drivers/multimodal.py:72]: After setting `self.position_by_modality = {'visual1': 1, 'visual2': 0, 'textual': 2}`, the following test will fail.
```
def test_multimodal_driver_with_shuffled_order(simple_multimodal_driver, mock_multimodal_encoder, doc_with_multimodal_chunks):
    simple_multimodal_driver.attach(executor=mock_multimodal_encoder, pea=None)
    simple_multimodal_driver._apply_all([doc_with_multimodal_chunks])
    doc = doc_with_multimodal_chunks
    assert len(doc.chunks) == 3
    visual1 = doc.chunks[1]
    visual2 = doc.chunks[0]
    textual = doc.chunks[2]
    control = np.concatenate([pb2array(visual2.embedding), pb2array(visual1.embedding), pb2array(textual.embedding)])
    test = pb2array(doc.embedding)
    np.testing.assert_array_equal(control, test)
```

**Review by nan-wang — CHANGES_REQUESTED** [review]

File: jina/drivers/multimodal.py
Comment (JoanFM) [reply jina/drivers/multimodal.py:54]: If you think about it, how else can you tell between two modalities by the executor? There must be some difference between every chunk, otherwise they must merge them in previous drivers.

File: jina/drivers/multimodal.py
Comment (JoanFM) [reply jina/drivers/multimodal.py:85]: I think the comment there is missplaced, what is needed for batching is the traversal logic.

File: jina/drivers/multimodal.py
Comment (JoanFM) [reply jina/drivers/multimodal.py:72]: will check!

Comment (JoanFM) [reply jina/drivers/multimodal.py]: Done Nan, added test, it was a changed we made trying to optimize, before was good

File: jina/drivers/multimodal.py
Comment (hanxiao) [inline jina/drivers/multimodal.py:21]: is this somehow related to https://github.com/jina-ai/jina/blob/521436c17a3a807bb47c020a4527b209651624c4/jina/drivers/helper.py#L90-L93

if yes, can we refactor it to unify them into one?

Comment (JoanFM) [reply jina/drivers/multimodal.py:21]: it is kind of the same, it differs I think in the way documents are passed. Since here we are looping on chunks internall. I thought about it but did not want to overcomplicate.

File: jina/drivers/multimodal.py
Comment (hanxiao) [inline jina/drivers/multimodal.py:30]:
```suggestion
        if modality in doc_content:
```

File: jina/drivers/multimodal.py
Comment (hanxiao) [inline jina/drivers/multimodal.py:26]: not sure i get this, will keep reading

File: jina/drivers/multimodal.py
Comment (hanxiao) [inline jina/drivers/multimodal.py:54]:
```suggestion
    .. note::
        - It traverses on the ``documents`` for which we want to apply the ``multimodal`` embedding. This way
        we can use the `batching` capabilities for the `executor`.
    .. warning::
```

File: jina/drivers/multimodal.py
Comment (hanxiao) [inline jina/drivers/multimodal.py:75]:
```suggestion
        for modality, val in self.position_by_modality.items():
            input_args[val] = content_by_modality[modality]
```

File: jina/drivers/multimodal.py
Comment (hanxiao) [inline jina/drivers/multimodal.py:63]:
```suggestion
    def position_by_modality(self) -> Dict[str, int]:
```

Comment (hanxiao) [reply jina/drivers/multimodal.py:63]: also need a docstring for this public property

File: jina/drivers/multimodal.py
Comment (hanxiao) [inline jina/drivers/multimodal.py:88]:
```suggestion
        num_modalities = len(self.position_by_modality)
```

File: jina/drivers/multimodal.py
Comment (hanxiao) [inline jina/drivers/multimodal.py:97]:
```suggestion
                for modality, val in doc_content.items():
                    content_by_modality[modality].append(val)
```

File: jina/drivers/multimodal.py
Comment (hanxiao) [inline jina/drivers/multimodal.py:108]: to me this part looks like a fancy `concat`, we do have a concat driver, not sure how they are related.

Comment (JoanFM) [reply jina/drivers/multimodal.py:108]: It is something that would be able to replace concatenating actually. You can see ConcatDriver as the combination of a `ReduceAll`plus a `MultiModalDriver` equipped with a `ConcatExecutor`

Comment (JoanFM) [reply jina/drivers/multimodal.py:108]: Concatenating is just another way of combining different modalities into one single vectorial representation. In the tests we have tested with such executors for instance

File: jina/executors/encoders/multimodal/__init__.py
Comment (hanxiao) [inline jina/executors/encoders/multimodal/__init__.py:2]: @bwanglzu btw, no need to add these two lines, they are added by our ci automatically

File: jina/executors/encoders/multimodal/__init__.py
Comment (hanxiao) [inline jina/executors/encoders/multimodal/__init__.py:31]:
```suggestion
```

This can be removed as decorator is not inherited in Python.

File: jina/drivers/multimodal.py
Comment (hanxiao) [inline jina/drivers/multimodal.py:31]: i dont get this early return, what's the purpose here @bwanglzu @JoanFM

Comment (JoanFM) [reply jina/drivers/multimodal.py:31]: This None will make it an invalid document similar to what happens in the EncodeDriver. It works under the assumption that every chunk should have different modality.

File: jina/drivers/multimodal.py
Comment (hanxiao) [reply jina/drivers/multimodal.py:75]: also, isn't a List[str] enough for `position_by_modality`?

Comment (hanxiao) [reply jina/drivers/multimodal.py:75]: if `position_by_modality` is a simple `List[str]` then `input_args` is simply a reorder of `content_by_modality` via:

```python
input_args = [content_by_modality[modality] for modality in position_by_modality]
```

Comment (JoanFM) [reply jina/drivers/multimodal.py:75]: yes it can be done, it is more a matter of taste, I like it more explicit but we can for for this

**Review by hanxiao — CHANGES_REQUESTED** [review]
i stop the review for now as I found `position_by_modality` can be a simple `List[str]`, no reason to make it a Dict

File: jina/drivers/multimodal.py
Comment (hanxiao) [reply jina/drivers/multimodal.py:75]: I went over the usage of `position_by_modality` and this really should be a list of str. This has to be fixed.

File: jina/drivers/multimodal.py
Comment (hanxiao) [reply jina/drivers/multimodal.py:31]: >  every chunk should have different modality

can you elaborate more on this? Say we have a PDF document, then the first granularity is
```
Doc
 |
 -- Chunk1 (image)
 |
 -- Chunk2 (text)
```
This assumption holds true only at this granularity, if we continue, this is not true

```
Doc
 |
 -- Chunk1 (image)
          |
          |-- ChunkChunk1 (image)
          |-- ChunkChunk2 (image)
 |
 -- Chunk2 (text)
```

File: jina/executors/encoders/multimodal/__init__.py
Comment (hanxiao) [inline jina/executors/encoders/multimodal/__init__.py:28]: is `self.position_by_modality` this really required at executor level? not seeing the executor using it, it is just used to inform the driver "please order in this way and give it to me". Perhaps just put `position_by_modality` directly into the driver?

File: tests/integration/multimodal/yaml/embedding-multimodal-concatenate.py
Comment (hanxiao) [inline tests/integration/multimodal/yaml/embedding-multimodal-concatenate.py:25]: i see, so it is a fancy concat delegated to the executor. so one can add fancy ops, e.g. MLP, PCA to it

File: jina/executors/encoders/multimodal/__init__.py
Comment (JoanFM) [reply jina/executors/encoders/multimodal/__init__.py:28]: It is not really required, I thought that the executor should have more knowledge than the driver on what is expected, but both are valid options.

File: jina/drivers/multimodal.py
Comment (hanxiao) [inline jina/drivers/multimodal.py:37]: hmm, now i see `MultiModalDriver` is a grouped version of `BaseEncodeDriver`, the grouping happens at `extract_docs`

File: jina/drivers/multimodal.py
Comment (JoanFM) [reply jina/drivers/multimodal.py:31]: it is true, but the user can have a previous multimodal driver to merge the `ChunkChunk1` and `ChunkChunk2` into `Chunk1`.

File: jina/executors/encoders/multimodal/__init__.py
Comment (bwanglzu) [reply jina/executors/encoders/multimodal/__init__.py:2]: ok, thanks :)

File: jina/executors/encoders/multimodal/__init__.py
Comment (JoanFM) [reply jina/executors/encoders/multimodal/__init__.py:28]: @bwanglzu I had a better thought and yes it makes more sense to have it at driver level. It is a quick change

File: jina/executors/encoders/multimodal/__init__.py
Comment (JoanFM) [inline jina/executors/encoders/multimodal/__init__.py:37]: Remove it from executor and move thr attribute to driver

File: jina/executors/encoders/multimodal/__init__.py
Comment (JoanFM) [inline jina/executors/encoders/multimodal/__init__.py:42]: This check made sense in driver but not much sense here.

File: tests/integration/multimodal/yaml/embedding-multimodal-concatenate.py
Comment (JoanFM) [inline tests/integration/multimodal/yaml/embedding-multimodal-concatenate.py]: Keep it here as it will be used in the future

File: tests/integration/multimodal/yaml/embedding-multimodal-concatenate.py
Comment (JoanFM) [inline tests/integration/multimodal/yaml/embedding-multimodal-concatenate.py]: Leep them here as they will be used in the future

Comment (JoanFM) [reply tests/integration/multimodal/yaml/embedding-multimodal-concatenate.py]: They did not make sense in the parent class but do here

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/executors/encoders/multimodal/__init__.py
Comment (bwanglzu) [reply jina/executors/encoders/multimodal/__init__.py:37]: ah I misunderstood.. I'll move it back

File: tests/integration/multimodal/yaml/embedding-multimodal-concatenate.py
Comment (bwanglzu) [reply tests/integration/multimodal/yaml/embedding-multimodal-concatenate.py]: ok

File: tests/integration/multimodal/yaml/embedding-multimodal-concatenate.py
Comment (bwanglzu) [reply tests/integration/multimodal/yaml/embedding-multimodal-concatenate.py]: ok

File: jina/drivers/multimodal.py
Comment (bwanglzu) [reply jina/drivers/multimodal.py]: updated

File: jina/drivers/multimodal.py
Comment (JoanFM) [inline jina/drivers/multimodal.py:64]: The point is to have it directly as an attribute of driver and not as a property reading from executor. So that the executor does not know anything about it

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/executors/encoders/multimodal/__init__.py
Comment (JoanFM) [reply jina/executors/encoders/multimodal/__init__.py:27]: @bwanglzu second thoughts. I would keep it as attribute of executor and access via property in the driver. This way the executor can use that to interpret the arguments of the encoders.

Comment (nan-wang) [reply jina/executors/encoders/multimodal/__init__.py:27]: What about grouping them into `dict` and passing into the executors?

Comment (JoanFM) [reply jina/executors/encoders/multimodal/__init__.py:27]: It makes a much worse interface. But also in my opinion mulimodal is about having N inputs that come from different  sources and that need N different treatments. It seems a fair assumption in my opinion

**Review by hanxiao — COMMENTED** [review]
rename `position_by_modality` -> `positional_modality`? analogy to `positional args` which is well accepted

**Comment by JoanFM** [comment]
> rename `position_by_modality` -> `positional_modality`? analogy to `positional args` which is well accepted

makes sense since is no longer a Dict

File: jina/drivers/multimodal.py
Comment (hanxiao) [inline jina/drivers/multimodal.py:63]: should be `positional_modality`

**Review by hanxiao — APPROVED** [review]

**Review by JoanFM — APPROVED** [review]

**Review by nan-wang — APPROVED** [review]

**Review by maximilianwerk — APPROVED** [review]

**Pull request merged**
